doc/memos/rdm-draft-alamos-lanzieri-runtime-configuration-architecture.md

```diff
@@ -0,0 +1,139 @@
+runtime configuration manager could in the future be replaced by a generic
+manager.  ![](./files/rdm-draft-alamos-lanzieri-runtime-configuration-architecture/architecture.png "Runtime Configuration Architecture")
+
+# 3. The RIOT Registry
+The RIOT Registry exposes an API for handling 
+**persistent module configurations**. The RIOT registry
+allows, among others, to:
+- Get or set configuration parameters via registry handlers
+- Commit a configuration group via the given registry handler
+- Load and set configuration parameters from a persistent storage device
+- Export configuration parameters (e.g copy to a buffer, print, etc)
+
+The RIOT registry consists of:
+- A list of Registry Handlers
+- A list of one or more Storage Facilities sources
+- One Storage destination
+
+# 4. Registry handlers
+A Registry Handler represents a configuration group in the RIOT Registry.
+Examples of configuration groups could be _Application_, _LoRaWAN keys_, etc.
+
+A Registry Handler is defined at least by a name and a series of handlers for interacting with the configuration parameters of the group. These handlers are:
+- `set`: Sets a value to a configuration parameter. This handler MUST take care
+of any logic to apply the value (e.g. data validation) or cache it until the commit handler is called
+- `get`: Gets the current value of a configuration parameter
+- `commit`: To be called when configuration parameters have been loaded from storage. It could be use for special needed logic on applying configuration parameters (e.g dependencies)
+- `export`: Calls an `export function` for each configuration parameter, with
+its name and value. Depending on the behavior of `export function`, this can be used for printing out all configurations, saving them in a persistent storage,
+etc.
+
+# 5. Storage facilities
+Storage facilities MUST implement the **storage
+interface** to allow the RIOT Registry to load, search and save configuration
+parameters. From the point of view of the RIOT Registry all parameters are
+key/value strings, it is responsibility of the storage facility to transform
+that to the proper format for storage (e.g. lines separated by `\n` character in
+a file).
+
+If a storage implements any kind of encryption mechanism it MUST be
+transparent to the RIOT Registry, and all that logic should be handled withing
+the facility.
+
+# 6. Configuration manager and interfaces
+The configuration manager is a pseudo-module that allows a RIOT node to be
+configured from one or more communication interfaces. Examples of these
+communication interfaces could be UART, SPI or higher layers like PPP, IPv6,
+UDP, CoAP, LWM2M, etc.
+
+The configuration manager MAY provide an access control mechanism for
+restricting access to the configurations. However, it can relay on security
+implemented on the layers mentioned above (e.g CoAP and DTLS, CHAP on PPP, etc)
```

Comment:
Do we want this? Or should we rather define a clear and robust access control model to the runtime configuration system of RIOT? (DTLS session keys may be obtained for reporting sensor readings to the infrastructure ...)

Reply:
Maybe lower layer security is just an extra and we still need access control. I know LWM2M already provides access control (and it's intended to be used as a Device Manager). In that case, the Configuration Manager is LWM2M + RIOT Registry. For other cases (CoAP), we might need to implement an access control mechanism in the endpoint.